File: abc212/e/a.cc
// https://atcoder.jp/contests/abc212/tasks/abc212_e

#include <bits/stdc++.h>
using namespace std;

int main() {
  const int MOD = 998244353;
  int N, M, K;
  cin >> N >> M >> K;
  vector<vector<long long> > graph(N);
  for (int m = 0; m < M; m++) {
    int U, V;
    cin >> U >> V;
    graph.at(U - 1).push_back(V - 1);
    graph.at(V - 1).push_back(U - 1);
  }
  vector<vector<long long> > dp(K + 1, vector<long long>(N));
  dp[0][0] = 1;
  for (int k = 1; k <= K; k++) {
    long long total = 0;
    for (int n = 0; n < N; n++) {
      total += dp[k - 1][n];
    }
    for (int n = 0; n < N; n++) {
      dp[k][n] = total - dp[k - 1][n];
      for (size_t i = 0; i < graph.at(n).size(); i++) {
	dp[k][n] -= dp[k - 1][graph.at(n).at(i)];
      }
      dp[k][n] %= MOD;
    }
  }
  cout << dp[K][0] << endl;
  return 0;
}

File: abc212/e/a.py
# https://atcoder.jp/contests/abc212/tasks/abc212_e

MOD = 998244353
N, M, K = map(int, input().split())
graph = [[] for n in range(N)]
for m in range(M):
    U, V = map(int, input().split())
    graph[U - 1].append(V - 1)
    graph[V - 1].append(U - 1)
dp = [[0] * N for k in range(K + 1)]
dp[0][0] = 1
for k in range(1, K + 1):
    total = 0
    for n in range(N):
        total += dp[k - 1][n]
    total %= MOD
    for n in range(N):
        dp[k][n] = total - dp[k - 1][n]
        dp[k][n] %= MOD
        for v in graph[n]:
            dp[k][n] -= dp[k - 1][v]
            dp[k][n] %= MOD
print(dp[K][0] % MOD)
